### PR TITLE
Multidevice Executor add accessor to fusion executor caches

### PR DESCRIPTION
**Pull request by cowanmeg**
Add `MultiDeviceExecutor::getFusionExecutorCaches` which returns a vector of pointers to the multidevice executor's fusion executor caches.

Note, Also renames the field `workspace` to `workspace_` to match naming conventions. 

cc @Priya2698 for https://github.com/NVIDIA/Fuser/pull/2386

### CODE REVIEW

**Comment by cowanmeg** [comment]
!build

File: csrc/multidevice/executor.cpp [hunk 1]
```diff
@@ -277,4 +277,18 @@ std::ostream& MultiDeviceExecutor::print() {
...
 }
 
+std::vector<FusionExecutorCache*> MultiDeviceExecutor::
+    getFusionExecutorCaches() {
+  NVF_CHECK(
+      params_.use_fusion_executor_cache,
+      "MultideviceExecutor must be configured to use FusionExecutorCache");
+  std::vector<FusionExecutorCache*> fecs;
+  for (SegmentedGroup* group : workspace_.group_run_order) {
+    if (!is_resharding_.at(group)) {
```
Comment (wujingyue) [inline csrc/multidevice/executor.cpp:287]: What do you think about
```suggestion
    if (fec_.count(group) > 0) {
```

**Review by samnordmann — COMMENTED** [review]
LGTM! One thing: I'm wondering how the caller of `getFusionExecutorCaches` can figure out what is the order of the `FusionExecutorCache` stored in the vector without accessing `workspace_. group_run_order`. Is my comment relevant?

**Comment by cowanmeg** [comment]
> LGTM! One thing: I'm wondering how the caller of `getFusionExecutorCaches` can figure out what is the order of the `FusionExecutorCache` stored in the vector without accessing `workspace_.group_run_order`. Is my comment relevant?

Yup relevant question. I think the only practical use case is for simple tests like these matmul tests where we have a linear sequence of computation so we know the order. The background for this PR is for matmul with ATen execution we want to check that the executor actually goes through ATen (https://github.com/NVIDIA/Fuser/pull/2386#issuecomment-2162265382) but to do that we need to access the FusionExecutorCache.

**Review by wujingyue — APPROVED** [review]

**Comment by Priya2698** [comment]
Thanks for this PR -- I added these checks to the multidevice matmul tests!

**Pull request merged**